### ISSUE TITLE
Regression in ACME cert acquisition reliability

### ISSUE THREAD

**Issue by consideRatio** [traefik/traefik#7784]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I upgraded from v2.3.7 to v2.4.0 leaving everything else the same, and started experiencing a significant amount of intermittent failures to acquire certificates as part of the associated test suite setup in the context where we use Traefik solely as a TLS termination proxy that also acquires ACME certs.

### What did you expect to see?

No increase in failures to acquire certs in a HTTP01 ACME based interaction.

### What did you see instead?

Intermittent failures

### Output of `traefik version`: (_What version of Traefik are you using?_)

Docker image v2.4.0.

```
Version:      2.4.0
Codename:     livarot
Go version:   go1.15.6
Built:        2021-01-19T17:26:51Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

I'm a maintainer of an open source project to deploy JupyterHub on k8s with a Helm chart ([jupyterhub/zero-to-jupyterhub-k8s](https://github.com/jupyterhub/zero-to-jupyterhub-k8s)). As part of that, we use Traefik as a TLS termination proxy. We have CI integration tests against a temporary k8s cluster with a ACME sever meant for testing purposes called pebble developed by the Let's encrypt team.

Below is the configuration files we provide Traefik.

```yaml
  # File number 1 - mounted to /etc/traefik/traefik.yaml
  traefik.yaml:
    accessLog:
      fields:
        headers:
          names:
            Authorization: redacted
            Cookie: redacted
            Set-Cookie: redacted
            X-Xsrftoken: redacted
      filters:
        statusCodes:
        - 500-599
    certificatesResolvers:
      default:
        acme:
          caServer: https://pebble/dir
          email: jovyan@jupyter.test
          httpChallenge:
            entryPoint: http
          storage: /etc/acme/acme.json
    entryPoints:
      http:
        address: :8080
      https:
        address: :8443
        transport:
          respondingTimeouts:
            idleTimeout: 10m0s
    log:
      level: DEBUG
    providers:
      file:
        filename: /etc/traefik/dynamic.yaml

  # File number 2 - mounted to /etc/traefik/dynamic.yaml
  dynamic.yaml:
    http:
      middlewares:
        hsts:
          headers:
            stsIncludeSubdomains: false
            stsPreload: false
            stsSeconds: 15724800
        redirect:
          redirectScheme:
            permanent: true
            scheme: https
        scheme:
          headers:
            customRequestHeaders:
              X-Scheme: https
      routers:
        default:
          entrypoints:
          - https
          middlewares:
          - hsts
          - scheme
          rule: PathPrefix(`/`)
          service: default
          tls:
            certResolver: default
            domains:
            - main: local.jovyan.org
            options: default
        insecure:
          entrypoints:
          - http
          middlewares:
          - redirect
          rule: PathPrefix(`/`)
          service: default
      services:
        default:
          loadBalancer:
            servers:
            - url: http://proxy-http:8000/
    tls:
      options:
        default:
          cipherSuites:
          - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
          - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
          - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
          - TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
          - TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
          - TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
          minVersion: VersionTLS12
          sniStrict: true
```

### Logs from an intermittent failure

__Traefik logs__

```
time="2021-01-20T17:36:09Z" level=debug msg="Building ACME client..." providerName=default.acme
time="2021-01-20T17:36:09Z" level=debug msg="https://pebble/dir" providerName=default.acme
time="2021-01-20T17:36:09Z" level=info msg=Register... providerName=default.acme
time="2021-01-20T17:36:09Z" level=debug msg="legolog: [INFO] acme: Registering account for jovyan@jupyter.test"
time="2021-01-20T17:36:09Z" level=debug msg="Using HTTP Challenge provider." providerName=default.acme
time="2021-01-20T17:36:09Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: Obtaining bundled SAN certificate"
time="2021-01-20T17:36:09Z" level=debug msg="legolog: [INFO] [local.jovyan.org] AuthURL: https://pebble/authZ/TQFlBqUT-Da0iCDBQp-Ru-F8dvT2JFgjdSBZ19fHdCo"
time="2021-01-20T17:36:09Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: Could not find solver for: tls-alpn-01"
time="2021-01-20T17:36:09Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: use http-01 solver"
time="2021-01-20T17:36:09Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: Trying to solve HTTP-01"
time="2021-01-20T17:36:09Z" level=debug msg="No default certificate, generating one"
time="2021-01-20T17:36:09Z" level=warning msg="No domain found in rule PathPrefix(`/`), the TLS options applied for this router will depend on the hostSNI of each request" entryPointName=https routerName=default@file
time="2021-01-20T17:36:09Z" level=debug msg="Looking for provided certificate(s) to validate [\"local.jovyan.org\"]..." providerName=default.acme
time="2021-01-20T17:36:09Z" level=debug msg="No ACME certificate generation required for domains [\"local.jovyan.org\"]." providerName=default.acme
time="2021-01-20T17:36:15Z" level=debug msg="legolog: [INFO] Deactivating auth: https://pebble/authZ/TQFlBqUT-Da0iCDBQp-Ru-F8dvT2JFgjdSBZ19fHdCo"
time="2021-01-20T17:36:15Z" level=error msg="Unable to obtain ACME certificate for domains \"local.jovyan.org\" : unable to generate a certificate for the domains [local.jovyan.org]: error: one or more domains had a problem:\n[local.jovyan.org] acme: error: 403 :: urn:ietf:params:acme:error:unauthorized :: Non-200 status code from HTTP: http://local.jovyan.org:80/.well-known/acme-challenge/vD3mbW1cdgc0O3j3rz4KbexOcDfqJi1Sy_xPvUvjpA8 returned 404, url: \n" providerName=default.acme
```

__Pebble logs (ACME server)__

```
Pebble 2021/01/20 17:35:07 ACME directory available at: https://:8443/dir
Pebble 2021/01/20 17:36:09 GET /dir -> calling handler()
Pebble 2021/01/20 17:36:09 HEAD /nonce-plz -> calling handler()
Pebble 2021/01/20 17:36:09 POST /sign-me-up -> calling handler()
Pebble 2021/01/20 17:36:09 There are now 1 accounts in memory
Pebble 2021/01/20 17:36:09 POST /order-plz -> calling handler()
Pebble 2021/01/20 17:36:09 There are now 1 authorizations in the db
Pebble 2021/01/20 17:36:09 Added order "yVhDh-EfnTVAiqAlxjNee8H3rwKO4X5ayasZnpwI0wY" to the db
Pebble 2021/01/20 17:36:09 There are now 1 orders in the db
Pebble 2021/01/20 17:36:09 POST /authZ/ -> calling handler()
Pebble 2021/01/20 17:36:09 POST /chalZ/ -> calling handler()
Pebble 2021/01/20 17:36:09 Pulled a task from the Tasks queue: &va.vaTask{Identifier:acme.Identifier{Type:"dns", Value:"local.jovyan.org"}, Challenge:(*core.Challenge)(0xc00008abe0), Account:(*core.Account)(0xc0000e8900)}
Pebble 2021/01/20 17:36:09 Starting 3 validations.
Pebble 2021/01/20 17:36:09 Attempting to validate w/ HTTP: http://local.jovyan.org:80/.well-known/acme-challenge/vD3mbW1cdgc0O3j3rz4KbexOcDfqJi1Sy_xPvUvjpA8
Pebble 2021/01/20 17:36:09 Attempting to validate w/ HTTP: http://local.jovyan.org:80/.well-known/acme-challenge/vD3mbW1cdgc0O3j3rz4KbexOcDfqJi1Sy_xPvUvjpA8
Pebble 2021/01/20 17:36:09 Attempting to validate w/ HTTP: http://local.jovyan.org:80/.well-known/acme-challenge/vD3mbW1cdgc0O3j3rz4KbexOcDfqJi1Sy_xPvUvjpA8
Pebble 2021/01/20 17:36:09 POST /authZ/ -> calling handler()
Pebble 2021/01/20 17:36:09 authz TQFlBqUT-Da0iCDBQp-Ru-F8dvT2JFgjdSBZ19fHdCo set INVALID by completed challenge I5Gj4EZMmCdTb-N8cLfyvXiGp5wWXadV9_88Cb6BX-g
Pebble 2021/01/20 17:36:09 order yVhDh-EfnTVAiqAlxjNee8H3rwKO4X5ayasZnpwI0wY set INVALID by invalid authz TQFlBqUT-Da0iCDBQp-Ru-F8dvT2JFgjdSBZ19fHdCo
Pebble 2021/01/20 17:36:15 POST /authZ/ -> calling handler()
Pebble 2021/01/20 17:36:15 POST /authZ/ -> calling handler()
Pebble 2021/01/20 17:36:15 POST /authZ/ -> calling handler()
```

### Logs from a successful attempt

__Traefik logs__

```
time="2021-01-20T18:30:27Z" level=debug msg="Building ACME client..." providerName=default.acme
time="2021-01-20T18:30:27Z" level=debug msg="https://pebble/dir" providerName=default.acme
time="2021-01-20T18:30:27Z" level=info msg=Register... providerName=default.acme
time="2021-01-20T18:30:27Z" level=debug msg="legolog: [INFO] acme: Registering account for jovyan@jupyter.test"
time="2021-01-20T18:30:27Z" level=debug msg="Using HTTP Challenge provider." providerName=default.acme
time="2021-01-20T18:30:27Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: Obtaining bundled SAN certificate"
time="2021-01-20T18:30:27Z" level=debug msg="legolog: [INFO] [local.jovyan.org] AuthURL: https://pebble/authZ/3fzV9XeMgBq0gRyWDrN3R8yBbKL8U6UxVrjhCYnSar8"
time="2021-01-20T18:30:27Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: Could not find solver for: tls-alpn-01"
time="2021-01-20T18:30:27Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: use http-01 solver"
time="2021-01-20T18:30:27Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: Trying to solve HTTP-01"
time="2021-01-20T18:30:27Z" level=debug msg="Retrieving the ACME challenge for token Mt7-7QH2LJxImRq-7cKWU6Fj40DX7HGekczlbSB0QdQ..." providerName=acme
time="2021-01-20T18:30:27Z" level=debug msg="Retrieving the ACME challenge for token Mt7-7QH2LJxImRq-7cKWU6Fj40DX7HGekczlbSB0QdQ..." providerName=acme
time="2021-01-20T18:30:27Z" level=debug msg="Retrieving the ACME challenge for token Mt7-7QH2LJxImRq-7cKWU6Fj40DX7HGekczlbSB0QdQ..." providerName=acme
time="2021-01-20T18:30:27Z" level=debug msg="legolog: [INFO] [local.jovyan.org] The server validated our request"
time="2021-01-20T18:30:27Z" level=debug msg="legolog: [INFO] [local.jovyan.org] acme: Validations succeeded; requesting certificates"
time="2021-01-20T18:30:29Z" level=debug msg="legolog: [INFO] Wait for certificate [timeout: 30s, interval: 500ms]"
time="2021-01-20T18:30:29Z" level=debug msg="legolog: [INFO] [local.jovyan.org] Server responded with a certificate."
time="2021-01-20T18:30:29Z" level=debug msg="Certificates obtained for domains [local.jovyan.org]" providerName=default.acme
time="2021-01-20T18:30:29Z" level=debug msg="Configuration received from provider default.acme: {\"http\":{},\"tls\":{}}" providerName=default.acme
time="2021-01-20T18:30:29Z" level=debug msg="Adding certificate for domain(s) local.jovyan.org"
time="2021-01-20T18:30:29Z" level=debug msg="No default certificate, generating one"
```

__Pebble logs (ACME server)__

```
Pebble 2021/01/20 18:29:05 ACME directory available at: https://:8443/dir
Pebble 2021/01/20 18:30:27 GET /dir -> calling handler()
Pebble 2021/01/20 18:30:27 HEAD /nonce-plz -> calling handler()
Pebble 2021/01/20 18:30:27 POST /sign-me-up -> calling handler()
Pebble 2021/01/20 18:30:27 There are now 1 accounts in memory
Pebble 2021/01/20 18:30:27 POST /order-plz -> calling handler()
Pebble 2021/01/20 18:30:27 There are now 1 authorizations in the db
Pebble 2021/01/20 18:30:27 Added order "c9WBWLePmrBYJeF0bmbwUMRx4Tk7IkAmLdnJHPkTC0s" to the db
Pebble 2021/01/20 18:30:27 There are now 1 orders in the db
Pebble 2021/01/20 18:30:27 POST /authZ/ -> calling handler()
Pebble 2021/01/20 18:30:27 POST /chalZ/ -> calling handler()
Pebble 2021/01/20 18:30:27 Pulled a task from the Tasks queue: &va.vaTask{Identifier:acme.Identifier{Type:"dns", Value:"local.jovyan.org"}, Challenge:(*core.Challenge)(0xc00033e280), Account:(*core.Account)(0xc00008cde0)}
Pebble 2021/01/20 18:30:27 Starting 3 validations.
Pebble 2021/01/20 18:30:27 Attempting to validate w/ HTTP: http://local.jovyan.org:80/.well-known/acme-challenge/Mt7-7QH2LJxImRq-7cKWU6Fj40DX7HGekczlbSB0QdQ
Pebble 2021/01/20 18:30:27 Attempting to validate w/ HTTP: http://local.jovyan.org:80/.well-known/acme-challenge/Mt7-7QH2LJxImRq-7cKWU6Fj40DX7HGekczlbSB0QdQ
Pebble 2021/01/20 18:30:27 Attempting to validate w/ HTTP: http://local.jovyan.org:80/.well-known/acme-challenge/Mt7-7QH2LJxImRq-7cKWU6Fj40DX7HGekczlbSB0QdQ
Pebble 2021/01/20 18:30:27 authz 3fzV9XeMgBq0gRyWDrN3R8yBbKL8U6UxVrjhCYnSar8 set VALID by completed challenge dKLTI2JV-g7bUwZO3g7Cpldzu8PwRemvD65Qmf2P1WM
Pebble 2021/01/20 18:30:27 POST /authZ/ -> calling handler()
Pebble 2021/01/20 18:30:29 POST /finalize-order/ -> calling handler()
Pebble 2021/01/20 18:30:29 Order c9WBWLePmrBYJeF0bmbwUMRx4Tk7IkAmLdnJHPkTC0s is fully authorized. Processing finalization
Pebble 2021/01/20 18:30:29 Issued certificate serial 1e1148e63484416f for order c9WBWLePmrBYJeF0bmbwUMRx4Tk7IkAmLdnJHPkTC0s
Pebble 2021/01/20 18:30:29 POST /my-order/ -> calling handler()
Pebble 2021/01/20 18:30:29 POST /certZ/ -> calling handler()
```

**Comment by jspdown** [comment]
Hello @consideRatio and thanks for your interest in Traefik. We haven't been able to reproduce the issue in the conditions you described. However, after a review of the changes introduced in the v2.4 and especially in https://github.com/traefik/traefik/pull/7458 we were able to force the system to behave like it did on your CI.

There's a concurrency issue between the configurations coming from the file provider and the internal provider. Because of this, there is a tiny chance that we start the challenge before Traefik is able to reply to pebble.

As a workaround, you can increase your timeout in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/6a3df2df1c7dc7cb2af52f71133945699694bb99/ci/common#L37 so Traefik dynamic configuration gets a second chance to converge. Alternatively you can replace your HTTP challenge by the TLS challenge which doesn't suffer from the same issue. It's not an ideal solution but at least it would prevent your CI from failing too often until we figure out a solution.

**Comment by consideRatio** [comment]
@jspdown thank you for your thorough investigation into this and work on traefik! :heart:

> There's a concurrency issue between the configurations coming from the file provider and the internal provider. Because of this, there is a tiny chance that we start the challenge before Traefik is able to reply to pebble.

Ah!

> As a workaround, you can increase your timeout in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/6a3df2df1c7dc7cb2af52f71133945699694bb99/ci/common#L37 so Traefik dynamic configuration gets a second chance to converge. 

Oh is traefik retrying after a while? If so, is this time interval defined somewhere?

Searching quickly for it myself, I found https://github.com/traefik/traefik/issues/2670 which indicate there won't be retries unless Traefik is manually restarted.

**Comment by jspdown** [comment]
@consideRatio We can't really talk about a real retry. Challenges will be (re-)attempted when a new dynamic configuration will be received on the ACME provider. 

In your case, here is what's happening:
- The File provider starts and emit a (partial) dynamic configuration
- The ACME provider starts and consumes this dynamic configuration. An acme certResolver is present in a router so we attempt to resolve the HTTP challenge. It fails because the Internal provider didn't have time to expose the HTTP challenge router.
- The Internal provider starts and emit a dynamic configuration (containing this time the HTTP challenge router)
- The ACME provider detects a change in the dynamic configuration and re-attempt the HTTP challenge, like if the fail never happened. In this case, the challenge is expected to succeed.

**Comment by consideRatio** [comment]
@jspdown what a beautiful breakdown of the events!

**Comment by traefiker** [comment]
Closed by #7805.
